lib/documents.js

```diff
@@ -2265,105 +2265,123 @@ function writeAllDocumentsImpl(jobOptions) {
-      if(operation.jobState.error){
-        summary = summary + 'Processing was cancelled by an error - '+operation.error.toString();
+
+      if(jobState.error){
+        summary = summary +'\n'+ 'Processing was cancelled by an error - '+jobState.error.toString();
       }
-      operation.jobState.jobOptions.onCompletion(summary);
+      summary = summary + '\n'+ 'timeElapsed = '+ jobState.timeElapsed+' milliseconds.';
+      jobState.jobOptions.onCompletion(summary);
     }
-
   }
 }
 
 function onWriteAllInit(output) {
 
-  let maxRequesters = output.length;
-  const operation = this;
-  const jobState = operation.jobState;
-
-  if(jobState.jobOptions && jobState.jobOptions.concurrentRequests) {
-    if(jobState.jobOptions.concurrentRequests.multipleOf &&
-        jobState.jobOptions.concurrentRequests.multipleOf === 'hosts'){
-      const hosts = new Set();
-      for(let i=0; i<output.length; i++) {
-        hosts.add(output[i].host);
-      }
-      maxRequesters = hosts.size;
-    }
-    const multiplier = (jobState.jobOptions.concurrentRequests.multiplier &&
-        jobState.jobOptions.concurrentRequests.multiplier>0)?
-        jobState.jobOptions.concurrentRequests.multiplier:0;
-    maxRequesters = (multiplier >0) ? (maxRequesters*multiplier):maxRequesters;
+  let maxRequesters = 0;
+  const jobState = this;
 
+  switch (jobState.jobOptions.concurrentRequests.multipleOf) {
+    case 'forests':
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier * output.length;
+      break;
+    case 'hosts':
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier *
+          new Set(output.map((forest) => forest.host)).size;
+      break;
+    default:
+      throw new Error('Invalid value for onCompletion.multipleOf. Value must be forests or hosts.');
   }
-
+  jobState.startTime = Date.now();
   for (let i = 0; i < maxRequesters; i++) {
     jobState.requesterCount++;
-    onWriteAllDocs(operation,i);
+    onWriteAllDocs(jobState,i);
   }
-
 }
 
-function onWriteAllDocs(operation, writerId) {
+function onWriteAllDocs(jobState, writerId) {
 
   const writeBatchArray = [];
-  const jobState = operation.jobState;
-  const batchSize = (jobState.jobOptions && jobState.jobOptions.batchSize &&
-      jobState.jobOptions.batchSize >0) ?
-      jobState.jobOptions.batchSize:100;
+  const batchSize = jobState.jobOptions.batchSize;
+  jobState.lastBatch = false;
+  jobState.retryCount = 2;
 
   a:
       for (let i = 0; i < batchSize; i++) {
         const record = jobState.stream.read();
 
-        if(jobState.jobOptions && jobState.jobOptions.defaultMetadata) {
-          writeBatchArray.push(jobState.jobOptions.defaultMetadata);
+        if(record !== null) {
+          if (jobState.jobOptions.defaultMetadata) {
+            writeBatchArray.push(jobState.jobOptions.defaultMetadata);
+          }
+          if (jobState.jobOptions.transform) {
+            writeBatchArray.push(jobState.jobOptions.transform);
```

Comment:
The transform should passed on the call to writeDocumentsImpl() -- see the JSDoc for writeDocuments()

It might be more convenient to use the object instead of the positional form of the call as in:

```
const batchdef = {documents: writeBatchArray};
if (jobState.jobOptions.transform) {
    batchdef.transform = jobState.jobOptions.transform;
}
writeDocumentsImpl.call(jobState.docInstance, false, batchdef)

```

Reply:
Done.